### DISCUSSION TITLE
Cross compiling of OSQP, FATROP, HPIPM for embedded use in dSPACE

### DISCUSSION THREAD

**Discussion by phmandl** [casadi/casadi#3884] [Q&A] [answered]
Hello,

i am currently on the journey to make several solvers available on the microautobox 3 or dSpace1403. I was already successful with OSQP, QRQP (easy thanks to CasADI) and FATROP. The idea is to have a full toolchain (casadi -> codegen -> auto sFunction with labels -> headers + libs for Space -> dSpace Configurationdesk) available soon which I would like to share with the community.  Something similar to ACADOS, however, is more user-friendly by being fully integrated into CASADI. 

However, for that, I have to cross-compile all solvers for the dSpace system. I faced some problems with the HPIPM solver.
Blasfeo + Fatrop works fine when I mex them; hpipm on the other hand not:

Error using mex
C:\Users\phman\repos\Z_casadi\libhpipm.a(d_ocp_qp.c.obj):d_ocp_qp.c:(.text+0x19c): undefined reference to
`blasfeo_memsize_dmat'
C:\Users\phman\repos\Z_casadi\libhpipm.a(d_ocp_qp.c.obj):d_ocp_qp.c:(.text+0x1b0): undefined reference to
`blasfeo_memsize_dmat'
C:\Users\phman\repos\Z_casadi\libhpipm.a(d_ocp_qp.c.obj):d_ocp_qp.c:(.text+0x1c5): undefined reference to
`blasfeo_memsize_dmat'
C:\Users\phman\repos\Z_casadi\libhpipm.a(d_ocp_qp.c.obj):d_ocp_qp.c:(.text+0x1d4): undefined reference to
`blasfeo_memsize_dvec'
C:\Users\phman\repos\Z_casadi\libhpipm.a(d_ocp_qp.c.obj):d_ocp_qp.c:(.text+0x249): undefined reference to
`blasfeo_memsize_dmat'
C:\Users\phman\repos\Z_casadi\libhpipm.a(d_ocp_qp.c.obj):d_ocp_qp.c:(.text+0x259): undefined reference to
`blasfeo_memsize_dmat'

I pulled hpipm and blasfeo from the same commits as casadi uses; e.g from the fork of Joris Gillis @ https://github.com/jgillis/hpipm.git and blasfeo as listed in the casadi cmak file.

Any idea why mex does not work? 
Help would be highly appreciated to further the development of this toolchain.

Best regards,
Philipp

**Comment by phmandl** [comment, marked as answer]
Nevermind...was quiet late.

I did not know that the mex command respects the order of the libs.

This works...
mex('-v', ['-I' inc_path],['-L' lib_path], '-lhpipm', '-lblasfeo', 'MPC.c')

This does not...
mex('-v', ['-I' inc_path],['-L' lib_path], '-lblasfeo', '-lhpipm', 'MPC.c')

Well, at least all are now cross-compiled.

**Reply by maskros-m** [reply]
Hi Philipp,

May I ask whether u managed to call .mex file in your simulink model when working with Fatrop?

Best regards,
Anh

**Reply by phmandl** [reply]
> Hi Philips,
> 
> May I ask whether u managed to call .mex file in your simulink model when working with Fatrop?
> 
> Best regards, Anh

You mean compiled on dSpace or just running a code-gen version of fatrop in an simulink s-function?

**Reply by maskros-m** [reply]
I am more interested in both steps, but currently struggling with calling .mex file in Simulink S-function. When running Simulink, I always get error: the .mex file is not found.

**Reply by phmandl** [reply]
To run fatrop in simulink, just follow this blog entry https://web.casadi.org/blog/s-function/
and compile with `mex('-v', ['-I' inc_path],  ['-L' lib_path], '-lblasfeo', '-lfatrop', 's_function.c', 'f.c')`

**Reply by phmandl** [reply]
How do you call your mex-file? You can't directly call the mex file as one does in matlab via scripting but rather only by converting it into a s-function (codegen) or with [coder.extrinsic(function)](https://de.mathworks.com/help/simulink/slref/coder.extrinsic.html#d126e415882) inside a matlab function which however is not codegen and therefore runs slower.

**Reply by maskros-m** [reply]
I actually followed the same routine and managed to generate .mex file with the aforementioned command. Including it to S-function block causes this issue, while running Simulink model. I am just curious whether u encounter the same issue?
![Unbenannt](https://github.com/user-attachments/assets/28ed3b9a-e9c3-4ebb-b4f4-dd1e28b81413)

**Reply by phmandl** [reply]
you have to be careful. you either modify the s_function.c template to use custom names or the function you compile with mex must be called f. Did you modify the s-function for your name `EOPT_fatrop_call.mexw64`

**Reply by phmandl** [reply]
If you use the template provided from the blog and want to use a nlp, I do the following to keep modifications simple: 

Codegen for the Van Der Pol example:

% BUILD CASADI SOLVER AS USUAL
opts.qpsol = 'osqp';
solver = casadi.nlpsol('solver', 'sqpmethod', prob, opts);

% CONSTRUCT FUNCTION OBJECT for MEX WITH SYMBOLICS
W0 = casadi.MX.sym('W0',size(w0));
LBW = casadi.MX.sym('LBW',size(lbw));
UBW = casadi.MX.sym('UBW',size(ubw));
LBG = casadi.MX.sym('LBG',size(lbg));
UBG = casadi.MX.sym('UBG',size(ubg));

fun_solver = solver('x0', W0, 'lbx', LBW, 'ubx', UBW, 'lbg', LBG, 'ubg', UBG);
fun_solver = casadi.Function('f',{W0,LBW,UBW,LBG,UBG},{fun_solver.x,fun_solver.f},{'w0','lbw','ubw','lbg','ubg'},{'x','f'});

cg_options = struct;
cg_options.casadi_real = 'real_T';
cg_options.casadi_int = 'int_T';
cg_options.with_header = true;
cg = casadi.CodeGenerator('f',cg_options);
cg.add_include('simstruc.h');
cg.add(fun_solver);
cg.generate();

lib_path = casadi.GlobalOptions.getCasadiPath();
inc_path = casadi.GlobalOptions.getCasadiIncludePath();

mex('-v', ['-I' inc_path],['-L' lib_path], '-losqp', 's_function.c', 'f.c')


Important here is that fun_solver is called 'f' for casadi as well as in the CodeGenerator command. This gives an f.h file for mex and the template s_function.c

I hope this is helpful.

**Reply by maskros-m** [reply]
Thanks Philipp, my problem is a little bit different, as my generated code is by Python. I will check the generated code
